tests/unit/lms/models/lti_user_test.py

```diff
@@ -68,3 +76,8 @@ def test_display_name(
     display_name_ = display_name(given_name, family_name, full_name)
 
     assert display_name_ == expected_display_name
+
+
+@pytest.fixture(autouse=True)
+def HUser(patch):
+    return patch("lms.models.lti_user.HUser")
```

Comment:
This test would change fairly radically if the above suggestions were in place, but I think this might be a good time to allow a deeper kind of testing. We can probably just check that we get the HUser object we expect, without mocking here.

Reply:
We actually didn't remove the mocks when we made the `display_name()` extraction with `LTIUser`, I just mocked `display_name()` instead:

https://github.com/hypothesis/lms/blob/b51a54d545e0dc1c7ef38a755e2778f9bf48c2a4/tests/unit/lms/validation/authentication/_launch_params_test.py#L11-L31

And I think for good reason. Take a look at how the `LTIUser.h_user` tests would look if they didn't mock `HUser` (if `username()` were extracted out of `HUser` and not mocked in `LTIUser.h_user`'s tests the results would be equivalent):

```diff
diff --git a/tests/unit/lms/models/lti_user_test.py b/tests/unit/lms/models/lti_user_test.py
index 2a60e4ce..f39a8054 100644
--- a/tests/unit/lms/models/lti_user_test.py
+++ b/tests/unit/lms/models/lti_user_test.py
-        HUser.from_lti_user.assert_called_once_with(lti_user)
-        assert h_user == HUser.from_lti_user.return_value
+        assert h_user == HUser(
+            username="2f42fed00de3039b0c720859d03b47",
+            display_name=lti_user.display_name,
+            provider=lti_user.tool_consumer_instance_guid,
+            provider_unique_id=lti_user.user_id,
+        )
```

The mock here isn't a problem in and of itself. For example, I don't think the tests without the mock are any deeper. Specifically, I don't think there's any way to break the code such that the with-mock tests would still be passing, whereas the without-mock tests would catch it. The with-mock tests provide complete coverage.

And the mock actually provides a bunch of advantages:

* The test "speaks" a lot more clearly about what `LTIUser.h_user` is actually responsible for doing, and what we're actually trying to test about it:

  ```python
   assert h_user == HUser.from_lti_user.return_value
   ```

   Compared to:

   ```python
   assert h_user == HUser(username="2f42fed00de3039b0c720859d03b47", ...)
   ```

   What the heck is this actually testing? What is `"2f42fed00de3039b0c720859d03b47"`? It's impossible to see from the test, but that opaque string is  SHA-1 hash of `tool_consumer_instance_guid` and `user_id`, which is an implementation detail of `HUser` that has nothing to do with `LTIUser` or to do with what this test is actually for.

* It also removes `display_name=lti_user.display_name`, `provider=lti_user.tool_consumer_instance_guid`, and `provider_unique_id=lti_user.user_id` from the `LTIUser` tests, all of which are `HUser` implementation details that have nothing to do with `LTIUser` or with what this test is supposed to be for

* It avoids duplicating tests, makes the tests less brittle, requires fewer test changes when you change the code. The without-mock version of the test above duplicates stuff that's already tested by `HUser.from_lti_user()`'s own unit tests. If you changed `HUser`'s internal implementation details, without the mock you'd have to update both `HUser` and `LTIUser`'s tests

* It means that `LTIUser`'s tests don't contribute to `HUser()`'s code coverage. `HUser` has its own unit tests and we want those to be responsible for its code coverage. We don't want misleadingly inflated coverage.

* It means that if there's a bug in `HUser.from_lti_user()` then `HUser.from_lti_user()`'s tests will be failing, but `LTIUser.h_user`'s tests will not be failing unless there's a bug in `LTIUser.h_user`. This is a good thing: a test should have only one reason to fail. This is easier to see if you imagine the pattern repeated again and again for the tests across an entire codebase. It's easier to pinpoint and debug test failures if only a few unit tests for the specific unit that has the bug are failing. If you frequently run the tests and get a lot of failing tests across a lot of different units, this wastes time trying to pinpoint where the bug is. Multiply the time lost again and again over the maintenance life of the codebase. Isolated tests scale better, they're easier to maintain.